Prototype selection API for shadow DOM

rniwa opened the pull request:
#### e29577b1a051a4c2cbec70904df6e2f7cf8b5102
<pre>
Prototype selection API for shadow DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=251997">https://bugs.webkit.org/show_bug.cgi?id=251997</a>

Reviewed by Wenson Hsieh.

Prototype Selection.prototype.getComposedRanges and Selection.prototype.direction as spec&apos;ed here:
<a href="https://w3c.github.io/selection-api/#dom-selection-getcomposedrange">https://w3c.github.io/selection-api/#dom-selection-getcomposedrange</a>
<a href="https://w3c.github.io/selection-api/#dom-selection-direction">https://w3c.github.io/selection-api/#dom-selection-direction</a>

One notable difference is that we support returning a sequence of StaticRanges as opposed to
a single range so the method name is accordingly getComposedRange&quot;s&quot;.

* LayoutTests/fast/shadow-dom/selection-direction-expected.txt: Added.
* LayoutTests/fast/shadow-dom/selection-direction.html: Added.
* LayoutTests/fast/shadow-dom/selection-getComposedRanges-expected.txt: Added.
* LayoutTests/fast/shadow-dom/selection-getComposedRanges.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/StaticRange.h:
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::direction const):
(WebCore::DOMSelection::setBaseAndExtent):
(WebCore::DOMSelection::getComposedRanges):
* Source/WebCore/page/DOMSelection.h:
* Source/WebCore/page/DOMSelection.idl:

Canonical link: <a href="https://commits.webkit.org/260121@main">https://commits.webkit.org/260121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06744e575457863194018dc4ab726cea5798c1c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16139 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116265 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110991 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7338 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99266 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112854 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40907 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27980 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82663 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96543 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9230 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29325 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95981 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7202 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9837 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6325 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15389 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48898 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104782 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11379 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25959 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3783 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/DOMSelection.cpp:
```diff
@@ -416,6 +433,40 @@ ExceptionOr<void> DOMSelection::removeRange(Range& liveRange)
+    if (!range)
+        return { };
+
+    HashSet<Ref<ShadowRoot>> shadowRootSet;
```
Any benefit to reserving initial capacity here?

rniwa replied:
Ah, good point. Will do.

rniwa committed:
Prototype selection API for shadow DOM
https://bugs.webkit.org/show_bug.cgi?id=251997

Reviewed by Wenson Hsieh.

Prototype Selection.prototype.getComposedRanges and Selection.prototype.direction as spec'ed here:
https://w3c.github.io/selection-api/#dom-selection-getcomposedrange
https://w3c.github.io/selection-api/#dom-selection-direction

One notable difference is that we support returning a sequence of StaticRanges as opposed to
a single range so the method name is accordingly getComposedRange"s".

* LayoutTests/fast/shadow-dom/selection-direction-expected.txt: Added.
* LayoutTests/fast/shadow-dom/selection-direction.html: Added.
* LayoutTests/fast/shadow-dom/selection-getComposedRanges-expected.txt: Added.
* LayoutTests/fast/shadow-dom/selection-getComposedRanges.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/StaticRange.h:
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::direction const):
(WebCore::DOMSelection::setBaseAndExtent):
(WebCore::DOMSelection::getComposedRanges):
* Source/WebCore/page/DOMSelection.h:
* Source/WebCore/page/DOMSelection.idl:

Canonical link: https://commits.webkit.org/260121@main
